src/util/memstream.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+//! Tokio compatible in-memory stream implementation. Helpful for testing.
+
+// TODO(povilas): move to future_utils crate - this module is generic and highly reusable.
+
+use priv_prelude::*;
+use std::fmt;
+use std::io::{self, Read, Write};
+
+/// In-memory stream implementing `AsyncWrite + AsyncRead`. This stream is not connected with
+/// any other stream. Meaning what you write to the stream you can later read from it.
+#[derive(Default)]
+pub struct EchoStream {
+    buf: Vec<u8>,
+    shutdown: bool,
+}
+
+impl Read for EchoStream {
+    /// Reads as much data as possible to the given buffer and removes that data from internal
+    /// buffer.
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if self.buf.is_empty() {
+            if self.shutdown {
+                return Ok(0);
+            } else {
+                return Err(io::ErrorKind::WouldBlock.into());
+            }
+        }
+
+        let mut bytes_read = 0;
+        while bytes_read < buf.len() && !self.buf.is_empty() {
+            buf[bytes_read] = self.buf.remove(0);
+            bytes_read += 1;
+        }
+        Ok(bytes_read)
+    }
+}
+
+impl Write for EchoStream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.buf.extend_from_slice(buf);
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl AsyncRead for EchoStream {}
+
+impl AsyncWrite for EchoStream {
+    /// Analogous to sending FIN to myself.
+    /// This will unblock `read()` calls which will get 0 as result value.
+    fn shutdown(&mut self) -> io::Result<Async<()>> {
+        self.shutdown = true;
+        Ok(Async::Ready(()))
+    }
+}
+
+impl fmt::Debug for EchoStream {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter
+            .debug_struct("EchoStream")
+            .field("buf", &self.buf)
+            .finish()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    mod memstream {
+        use super::*;
+        use tokio_core::reactor::Core;
+        use tokio_io;
+
+        mod read {
+            use super::*;
+
+            mod when_stream_is_empty {
+                use super::*;
+
+                #[test]
+                fn when_stream_is_open_it_returns_would_block_error() {
+                    let mut stream = EchoStream::default();
+
+                    let mut data = [0; 4096];
```

Comment:
Maybe make it much lesser in size to not have a big mem footprint in test cases if possible - like here it does not really matter what size it is because we are asserting an `EWOULDBLOCK`
Similarly other places in the test code - nothing incorrect, just wasteful.